module/zfs/sa.c

```diff
@@ -1285,7 +1285,16 @@ sa_build_index(sa_handle_t *hdl, sa_buf_type_t buftype)
 	/* only check if not old znode */
 	if (IS_SA_BONUSTYPE(bonustype) && sa_hdr_phys->sa_magic != SA_MAGIC &&
 	    sa_hdr_phys->sa_magic != 0) {
-		VERIFY(BSWAP_32(sa_hdr_phys->sa_magic) == SA_MAGIC);
+		if (BSWAP_32(sa_hdr_phys->sa_magic) != SA_MAGIC) {
+			mutex_exit(&sa->sa_lock);
+			cmn_err(CE_WARN, "Buffer Header: %x != SA_MAGIC:%x",
+				sa_hdr_phys->sa_magic, SA_MAGIC);
```

Comment:
I'm not that familiar with the rules around cmn_err(WARN) on linux.  Would it make sense to also (or instead) do a zfs_dbgmsg()?

Reply:
cmn_err() is provided by SPL (See [spl/spl-err.c](https://github.com/zfsonlinux/spl/blob/master/module/spl/spl-err.c#L94-L132)).
Basically it boils down to printk with KERN_WARNING.  zfs_dbgmsg() only prints if debugging is enabled. I looked at zfs_dbgmsg() initially, but I figured we'd always want to print, because this is evidence of data corruption.

Reply:
FYI, zfs_dbgmsg() now records the message by default, and you can retrieve it from /proc/spl/kstat/zfs/dbgmsg.  (And AFAIK it never prints to console.)

Reply:
I'll call zfs_dbgmsg() also but it looks like zfs_dbgmsg() can be disabled via modparam (though it is enabled by default).
https://github.com/zfsonlinux/zfs/blob/bc8a6a60e9f0fd219e10f355384d87a41d4f5882/include/sys/zfs_debug.h#L60-L62